### PR TITLE
feat:[CDS-70594]: supported primaryArtifactRef as expression

### PR DESCRIPTION
**Pull request by naman-goenka**
## Describe your changes
- Issue: 
a. currently when primaryArtifactRef is provided as expression in service, then serviceMetadata call in pipeline fails stating expressions are not allowed, but we can eventually run such a pipeline.
b. When the field is runtime input in service itself, but provided as expression in pipeline, the reconciliation fails.
- supported two cases related to expressions of primaryArtifactRef
a. When the fields is expression in service itself. ServiceMetadata call removes the `artifacts` block, and any manually added inputs for artifacts are considered in reconciliation. Provided, the expression for the field provided in yaml should be same as defined in service.
b. When the field is runtime input in service itself, but provided as expression in pipeline. ServiceMetadata call gives primaryArtifactRef as runtime input. Reconciliation preserves the provided expression, along with artifacts inputs.
- Testing: 
1. Testing a and b scenarios with a service containing 2 artifacts and runtime manifests. (Checked metadata call, reconciliation, execution, fixed/runtime manual inputs)
2. When in case a, different expression is provided in pipeline yaml then after reconciliation artifact block is reverted back.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49490)

### CODE REVIEW

**Comment by tchaurasiya** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java` line 178 at r1](https://reviewable.io/reviews/harness/harness-core/49490#-NYXTbBjDHaYTSvzxLSu:-NYXTbBjDHaYTSvzxLSv:b1ik3zj) ([raw file](https://github.com/harness/harness-core/blob/3af54c315f1a3ea61b255aea6dafd75891f06f1c/120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java#L178)):*
> ```Java
> 
>     YamlNode primaryArtifactRefNodeInServiceEntity = YamlNodeUtils.goToPathUsingFqn(
>         validateAndGetYamlNode(serviceYaml), "service.serviceDefinition.spec.artifacts.primary.primaryArtifactRef");
> ```

will this works for multi service case?

**Comment by tchaurasiya** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java` line 188 at r1](https://reviewable.io/reviews/harness/harness-core/49490#-NYXZ_0B097r1-QKXlLL:-NYXZ_0B097r1-QKXlLM:b-zbrjax) ([raw file](https://github.com/harness/harness-core/blob/3af54c315f1a3ea61b255aea6dafd75891f06f1c/120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java#L188)):*
> ```Java
>         && !NGExpressionUtils.isRuntimeField(primaryArtifactRefNodeValue)
>         && NGExpressionUtils.isExpressionField(primaryArtifactRefNodeValue)
>         && primaryArtifactRefNodeValueInServiceEntity.equals(primaryArtifactRefNodeValue)) {
> ```

can we use smaller methods here. Difficult to read.

File: 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java
Comment (hingerabhinav) [inline 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java:1]: Let's rename this to CDInputsValidationHelper

File: 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java [hunk 1]
```diff
@@ -169,14 +169,47 @@ void validateServiceInputs(
...
+        serviceYaml, serviceRef, primaryArtifactRefNodeValue);
+
+    YamlNode primaryArtifactRefNodeInServiceEntity = YamlNodeUtils.goToPathUsingFqn(
+        validateAndGetYamlNode(serviceYaml), "service.serviceDefinition.spec.artifacts.primary.primaryArtifactRef");
+    String primaryArtifactRefNodeValueInServiceEntity =
+        primaryArtifactRefNodeInServiceEntity == null ? null : primaryArtifactRefNodeInServiceEntity.asText();
+
+    if (EmptyPredicate.isNotEmpty(primaryArtifactRefNodeValueInServiceEntity)
+        && !NGExpressionUtils.isRuntimeField(primaryArtifactRefNodeValueInServiceEntity)
+        && NGExpressionUtils.isExpressionField(primaryArtifactRefNodeValueInServiceEntity)
```
Comment (naman-goenka) [inline 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java:184]: add a util of strictly expression field

File: 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java [hunk 2]
```diff
@@ -169,14 +169,47 @@ void validateServiceInputs(
...
+        && EmptyPredicate.isNotEmpty(primaryArtifactRefNodeValue)
+        && !NGExpressionUtils.isRuntimeField(primaryArtifactRefNodeValue)
+        && NGExpressionUtils.isExpressionField(primaryArtifactRefNodeValue)
+        && primaryArtifactRefNodeValueInServiceEntity.equals(primaryArtifactRefNodeValue)) {
+      // in case the primary artifact field is strictly expression in service entity
```
Comment (naman-goenka) [inline 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java:189]: move comment to top

Comment (naman-goenka) [reply 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java:189]: done

File: 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java (hunk 1 above)
Comment (naman-goenka) [reply 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java:184]: done

**Comment by naman-goenka** [comment]
> _[`120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java` line 188 at r1](https://reviewable.io/reviews/harness/harness-core/49490#-NYXZ_0B097r1-QKXlLL:-NYXZ_0B097r1-QKXlLM:b-zbrjax) ([raw file](https://github.com/harness/harness-core/blob/3af54c315f1a3ea61b255aea6dafd75891f06f1c/120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java#L188)):_
> 
> > ```java
> >         && !NGExpressionUtils.isRuntimeField(primaryArtifactRefNodeValue)
> >         && NGExpressionUtils.isExpressionField(primaryArtifactRefNodeValue)
> >         && primaryArtifactRefNodeValueInServiceEntity.equals(primaryArtifactRefNodeValue)) {
> > ```
> 
> can we use smaller methods here. Difficult to read.

done

File: 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java
Comment (naman-goenka) [reply 120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java:1]: done

**Comment by naman-goenka** [comment]
> _[`120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java` line 178 at r1](https://reviewable.io/reviews/harness/harness-core/49490#-NYXTbBjDHaYTSvzxLSu:-NYXTbBjDHaYTSvzxLSv:b1ik3zj) ([raw file](https://github.com/harness/harness-core/blob/3af54c315f1a3ea61b255aea6dafd75891f06f1c/120-ng-manager/src/main/java/io/harness/ng/core/refresh/helper/InputsValidationHelper.java#L178)):_
> 
> > ```java
> >     YamlNode primaryArtifactRefNodeInServiceEntity = YamlNodeUtils.goToPathUsingFqn(
> >         validateAndGetYamlNode(serviceYaml), "service.serviceDefinition.spec.artifacts.primary.primaryArtifactRef");
> > ```
> 
> will this works for multi service case?

yes , tested it.

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger uts

**Review by tchaurasiya — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49490)* status: 12 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by hingerabhinav — APPROVED** [review]
:lgtm:

> Reviewed 5 of 6 files at r2, 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49490#-:-NZ4MbPW4u0CyZdZnL9m:bnfp4nl)* status: 1. unresolved discussions.

**Comment by naman-goenka** [comment]
trigger review

**Pull request merged**